vkubelet/podcontroller.go

```diff
@@ -0,0 +1,322 @@
+		recorder:     recorder,
+		context:      context,
+	}
+
+	// Set up event handlers for when Pod resources change.
+	pc.podsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(pod interface{}) {
+			if key, err := cache.MetaNamespaceKeyFunc(pod); err != nil {
+				log.G(pc.context).Error(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Create a copy of the old and new pod objects so we don't mutate the cache.
+			oldPod := oldObj.(*corev1.Pod).DeepCopy()
+			newPod := newObj.(*corev1.Pod).DeepCopy()
+			// We want to check if the two objects differ in anything other than their resource versions.
+			// Hence, we make them equal so that this change isn't picked up by reflect.DeepEqual.
+			newPod.ResourceVersion = oldPod.ResourceVersion
+			// Skip adding this pod's key to the work queue if its .metadata (except .metadata.resourceVersion) and .spec fields haven't changed.
+			// This guarantees that we don't attempt to sync the pod every time its .status field is updated.
+			if reflect.DeepEqual(oldPod.ObjectMeta, newPod.ObjectMeta) && reflect.DeepEqual(oldPod.Spec, newPod.Spec) {
+				return
+			}
+			// At this point we know that something in .metadata or .spec has changed, so we must proceed to sync the pod.
+			if key, err := cache.MetaNamespaceKeyFunc(newPod); err != nil {
+				log.G(pc.context).Error(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+		DeleteFunc: func(pod interface{}) {
+			if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(pod); err != nil {
+				log.G(pc.context).Error(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	// Return the instance of PodController back to the caller.
+	return pc
+}
+
+// Run will set up the event handlers for types we are interested in, as well as syncing informer caches and starting workers.
+// It will block until stopCh is closed, at which point it will shutdown the work queue and wait for workers to finish processing their current work items.
+func (pc *PodController) Run(threadiness int) error {
+	defer runtime.HandleCrash()
+	defer pc.workqueue.ShutDown()
+
+	// Wait for the caches to be synced before starting workers.
+	if ok := cache.WaitForCacheSync(pc.context.Done(), pc.podsInformer.Informer().HasSynced); !ok {
+		return pkgerrors.New("failed to wait for caches to sync")
+	}
+
+	// Perform a reconciliation step that deletes any dangling pods from the provider.
+	// This happens only when the virtual-kubelet is starting, and operates on a "best-effort" basis.
+	// If by any reason the provider fails to delete a dangling pod, it will stay in the provider and deletion won't be retried.
+	pc.deleteDanglingPods()
+
+	// Launch two workers to process Pod resources.
+	log.G(pc.context).Info("starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(pc.runWorker, time.Second, pc.context.Done())
```

Comment:
It will still cause pod leaking if threadiness > 1!

If the user create a pod and delete it immediately, you will receive at least 2 events: ADD pod, DELETE pod. Both of them are en-queued. If there are 2 workers, they will probably be run in different thread. There is a chance that DELETE event is handled before ADD event, and cause the leaking of the pod in the provider.

Reply:
Actually, it won't.

To start with, we don't enqueue events anymore. We enqueue (pod) keys.

Workers popping from the queue are not aware of the type of event anymore. They'll always look into the state of both worlds (Kubernetes and provider) before making any decisions.

The work queue also offers us the guarantees we need. It guarantees, for example, that items are processed in the order in which they are queued. Furthermore, while a (pod) key is being processed by a worker, no other worker can process that same key - it is remembered by the work queue as being in "processing" state. That being, what would happen in the scenario you describe is roughly the following:

1. `ADDED` event causes `foo` to be added to the work queue.
2. `MODIFIED`/`DELETED` event causes `foo` to be added to the work queue.
3. Worker 1 pops `foo` to process. Processing takes N time. Worker 2 cannot pick `foo` to process in the meantime.
4. Worker 1 finishes processing. Since the key has been re-added in the meantime, it can now be picked by either Worker 1 or Worker 2. The process repeats from 3.

Alternatively, this can happen:

1. `ADDED` event causes `foo` to be added to the work queue.
2. Worker 1 immediately pops `foo` to process. Processing takes N time.
3. Roughly at the same time, `MODIFIED`/`DELETED` event causes `foo` to be added to the work queue. One of two things happen:
   1. `foo` is still being processed by Worker 1, in which case it is not added to the queue.
   2. `foo` has already been processed by Worker 1, in which case it is re-added and the process repeats itself.

For further information about the guarantees provided by the work queue, you can have a look at https://github.com/kubernetes/client-go/blob/master/util/workqueue/doc.go#L17-L25.

Reply:
[I've added tracing](https://github.com/virtual-kubelet/virtual-kubelet/pull/425#discussion_r237898598) so you can try and reproduce this behavior on your side, if you want to.